### PR TITLE
AMS-28: User should be able to create, read, update, and delete an asset

### PR DESCRIPTION
**Pull request by minega25**
### Description
As a user, I should be able to add, delete, get all and single asset.

### How to test
Test the following mutations and queries:
 - addAsset
 - allAsset
 - updateAsset
 - deleteAsset
 - asset

### How has this been tested?

[x] Unit testing
[ ] Integration testing
[ ] End-to-end testing

### Jira

[Delivers AMS-28]

### CODE REVIEW

**Comment by habinezadalvan** [comment]
Resolve the failing checks.

File: package.json [hunk 1]
```diff
@@ -58,7 +59,10 @@
     "ts-node": "^8.6.2"
   },
   "lint-staged": {
-      "src/**/*.(ts|js|tsx|jsx)": ["yarn lint", "git add"]
```
Comment (habinezadalvan) [inline package.json]: why did you remove the lint-staged check?

File: src/sequelize/seeders/5_user_seed.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+module.exports = {
+  up: (queryInterface) => queryInterface.bulkInsert('Users', [{
+    firstName: 'foo',
+    lastName: 'bar',
+    userName: 'foobar',
+    email: 'foobar@gmail.com',
+    password: 'sdfssgggg',
```
Comment (habinezadalvan) [inline src/sequelize/seeders/5_user_seed.js]: I think the password needs to be hashed, and there are some missing fields like phone number and gender. please check the other sheet with all table to see all required fields.

File: .sequelizerc [hunk 3]
```diff
@@ -4,5 +4,6 @@ module.exports={
     "config": path.resolve('./src/sequelize/config', 'config.js'),
     "models-path": path.resolve('./src/sequelize/models'),
     "seeders-path": path.resolve('./src/sequelize/seeders'),
+    "migrations-source-path": path.resolve('./src/sequelize/migrations'),
```
Comment (habinezadalvan) [inline .sequelizerc:7]: what is this for?

**Review by habinezadalvan — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (minega25) [reply package.json]: No I did not eslint just reformarted them. Check the actual file

File: src/sequelize/seeders/5_user_seed.js (hunk 2 above)
Comment (minega25) [reply src/sequelize/seeders/5_user_seed.js]: For some reason they are not required. But I just wanted the userId. You can create a seed that has all the data you need

File: .sequelizerc (hunk 3 above)
Comment (minega25) [reply .sequelizerc:7]: Its needed by sequelize cli for you to use the seed:generate command

**Comment by minega25** [comment]
> Resolve the failing checks.
This PR is still in progress. I just raised for you guys to get a direction to create your own seeds just as requested by Musinda

File: package.json [hunk 4]
```diff
@@ -9,30 +9,34 @@
     "lint": "eslint . --ext .js,.jsx,.ts,.tsx",
     "build": "yarn build-ts && yarn lint",
     "dev:watch": "nodemon -r ts-node src/index.ts",
-    "test": "NODE_ENV=test jest --coverage",
+    "test": "jest --coverage",
+    "pretest": "NODE_ENV=test yarn db:migrate:undo && NODE_ENV=test yarn db:migrate && NODE_ENV=test yarn db:seed:undo && NODE_ENV=test yarn db:seed",
```
Comment (habinezadalvan) [inline package.json]: What happens when you only write a single NODE_ENV=test from the beginning of the script? Instead of repeating it before each command.

File: src/sequelize/seeders/6_user_seed.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const bcrypt = require('bcrypt');
+require('dotenv/config');
+
+const salt = bcrypt.genSaltSync(10);
+const { SEEDER_ADMIN_PASSWORD, SEEDER_SUPER_ADMIN_PASSWORD } = process.env;
+const adminPassword = bcrypt.hashSync(SEEDER_ADMIN_PASSWORD, salt);
```
Comment (habinezadalvan) [inline src/sequelize/seeders/6_user_seed.js]: what do you think by adding @harerakalex 's utils file and you use his reusable functions for hashing the password? 
This is optional, you can even leave it like that then someone who will rebase after his PR got merged can change this.

File: src/services/asset.services.ts [hunk 6]
```diff
@@ -1,8 +1,61 @@
...
+    }
+    asset = await Asset.findOne({ where: { serialNo: id } });
+    await Redis.save(`asset:${id}`, asset);
+
+    return asset;
+  }
+
+  static async addAsset(asset): Promise<Asset> {
+    const assetSequelizeObj = await Asset.create(asset);
+    const validatedAsset = await assetSequelizeObj.save();
```
Comment (habinezadalvan) [inline src/services/asset.services.ts]: how do you validate asset user inputs? I didn't see any validation function or package for validation like joi or validatorjs.

**Review by habinezadalvan — COMMENTED** [review]
Wow, nice work @minega25. I like your algorithm and logic. Anyway, I left a few comments, you can check on the before your PR got merged.

File: src/services/asset.services.ts [hunk 7]
```diff
@@ -1,8 +1,61 @@
...
+
+    return allAssets;
+  }
+
+  static async getAssetBySerialNo(id): Promise<Asset | null> {
+    let asset = await Redis.fetch(`asset:${id}`);
+    if (asset) {
+      return asset;
+    }
+    asset = await Asset.findOne({ where: { serialNo: id } });
```
Comment (habinezadalvan) [inline src/services/asset.services.ts:21]: What happens in case that asset is not available? What do you think by validating such a scenario?

File: src/services/asset.services.ts [hunk 8]
```diff
@@ -1,8 +1,61 @@
...
+    await Redis.save(`asset:${id}`, asset);
+
+    return asset;
+  }
+
+  static async addAsset(asset): Promise<Asset> {
+    const assetSequelizeObj = await Asset.create(asset);
```
Comment (habinezadalvan) [inline src/services/asset.services.ts:35]: What happens in case the asset is already in the system? What do you think by first checking if an asset with a certain serialNo is already in the system?

**Review by habinezadalvan — CHANGES_REQUESTED** [review]
validate if the asset is already in the system before you create it and send a error in case you want to fetch an asset that is not in the system.

File: src/sequelize/seeders/6_user_seed.js (hunk 5 above)
Comment (minega25) [reply src/sequelize/seeders/6_user_seed.js]: Since its impossible for us to use typescript or es6 in sequelize seeders, its not possible to export functions and import them in es5 modules in this case. Hence I am leaving this as it is. Unless there is another solution I am not aware of

**Review by habinezadalvan — APPROVED** [review]
LGTM

**Review by harerakalex — APPROVED** [review]
Great job @minega25, LGTM
